File: src/types/telemetry.ts
/**
 * Данные для 4 колёс [FL, FR, RL, RR]
 */
export type WheelData = [number, number, number, number];

/**
 * Единичный sample телеметрии (нормализованный)
 * Все данные интерполированы к единому timestamp
 */
export interface TelemetrySample {
  // Время и позиция
  timestamp: number; // seconds from lap start
  lapDistance: number; // meters from lap start
  totalDistance: number; // meters from session start

  // Позиция в пространстве (для отрисовки трека)
  posX: number; // GPS Latitude (relative coords)
  posY: number; // GPS Longitude (relative coords)

  // Скорость
  speed: number; // km/h (Ground Speed)
  wheelSpeed: WheelData; // m/s

  // Inputs водителя
  throttle: number; // 0-100 %
  brake: number; // 0-100 %
  clutch: number; // 0-100 %
  steering: number; // degrees or normalized

  // Двигатель
  rpm: number;
  gear: number; // -1 = reverse, 0 = neutral, 1-8 = gears

  // G-силы
  gForceX: number; // lateral (G)
  gForceY: number; // longitudinal (G)
  gForceZ: number; // vertical (G)

  // Шины - температуры (°C)
  tireTempLeft: WheelData;
  tireTempCenter: WheelData;
  tireTempRight: WheelData;

  // Шины - другие параметры
  tirePressure: WheelData; // kPa
  tireWear: WheelData; // % remaining

  // Тормоза
  brakeTemp: WheelData; // °C

  // Топливо
  fuelLevel: number; // liters

  // Подвеска
  suspensionPos: WheelData; // m
  rideHeightFront: number; // m
  rideHeightRear: number; // m
}

/**
 * Облегчённые данные для отрисовки траектории
 */
export interface TrajectoryPoint {
  x: number; // normalized or GPS latitude
  y: number; // normalized or GPS longitude
  distance: number; // lap distance (m)
  speed: number; // km/h
  throttle?: number; // optional for coloring
  brake?: number; // optional for coloring
}

/**
 * Данные телеметрии для одного круга
 */
export interface LapTelemetry {
  lapNumber: number;
  samples: TelemetrySample[];
  sampleRate: number; // effective Hz after normalization
  duration: number; // seconds
  distance: number; // meters
}

/**
 * Минимальные данные для графиков (один канал)
 */
export interface ChannelData {
  timestamps: number[]; // or distances
  values: number[];
}

/**
 * Данные для графика с несколькими каналами
 */
export interface ChartData {
  xAxis: 'time' | 'distance';
  xValues: number[];
  channels: {
    name: string;
    values: number[];
    unit: string;
    color?: string;
  }[];
}

File: src-tauri/src/models/telemetry.rs
use serde::{Deserialize, Serialize};

/// Данные для 4 колёс [FL, FR, RL, RR]
pub type WheelData = [f32; 4];

/// Единичный sample телеметрии (нормализованный)
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct TelemetrySample {
    // Время и позиция
    pub timestamp: f64,
    pub lap_distance: f32,
    pub total_distance: f32,

    // Позиция в пространстве
    pub pos_x: f32,
    pub pos_y: f32,

    // Скорость
    pub speed: f32,
    pub wheel_speed: WheelData,

    // Inputs
    pub throttle: f32,
    pub brake: f32,
    pub clutch: f32,
    pub steering: f32,

    // Двигатель
    pub rpm: f32,
    pub gear: i8,

    // G-силы
    pub g_force_x: f32,
    pub g_force_y: f32,
    pub g_force_z: f32,

    // Шины - температуры
    pub tire_temp_left: WheelData,
    pub tire_temp_center: WheelData,
    pub tire_temp_right: WheelData,

    // Шины - другое
    pub tire_pressure: WheelData,
    pub tire_wear: WheelData,

    // Тормоза
    pub brake_temp: WheelData,

    // Топливо
    pub fuel_level: f32,

    // Подвеска
    pub suspension_pos: WheelData,
    pub ride_height_front: f32,
    pub ride_height_rear: f32,
}

/// Облегчённые данные для отрисовки траектории
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TrajectoryPoint {
    pub x: f32,
    pub y: f32,
    pub distance: f32,
    pub speed: f32,
    pub throttle: Option<f32>,
    pub brake: Option<f32>,
}

/// Данные телеметрии для одного круга
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LapTelemetry {
    pub lap_number: u32,
    pub samples: Vec<TelemetrySample>,
    pub sample_rate: f32,
    pub duration: f64,
    pub distance: f32,
}

impl Default for LapTelemetry {
    fn default() -> Self {
        Self {
            lap_number: 0,
            samples: Vec::new(),
            sample_rate: 10.0,
            duration: 0.0,
            distance: 0.0,
        }
    }
}

/// Точка границы трека (envelope из всех телеметрий)
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TrackBoundaryPoint {
    /// Дистанция вдоль трека
    pub distance: f32,
    /// Минимальные координаты (самая внутренняя точка)
    pub min_x: f32,
    pub min_y: f32,
    /// Максимальные координаты (самая внешняя точка)
    pub max_x: f32,
    pub max_y: f32,
    /// Центральная линия (среднее всех траекторий)
    pub center_x: f32,
    pub center_y: f32,
    /// Количество проходов через эту точку
    pub sample_count: u32,
}

/// Envelope границ трека на основе всех телеметрий
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TrackBoundaryEnvelope {
    /// Имя трека
    pub track_name: String,
    /// Общая длина трека
    pub track_length: f32,
    /// Точки границ
    pub points: Vec<TrackBoundaryPoint>,
    /// Количество проанализированных кругов
    pub total_laps: u32,
    /// Количество проанализированных файлов
    pub total_files: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Corner {
    pub id: u32,
    pub name: String,
    pub start_index: usize,
    pub end_index: usize,
    pub apex_index: usize,
    pub min_speed: f32,
    pub entry_speed: f32,
    pub exit_speed: f32,
    pub turn_direction: String, // "left" or "right"
    pub max_curvature: f32,
}
